R/joint_optspace_helper.R

```diff
@@ -0,0 +1,105 @@
+#' Joint RPCA Ordination Across Multiple Compositional Tables
+#'
+#' Internal function that performs Robust PCA via joint OptSpace decomposition across multiple compositional tables.
+#' It splits each table into train/test sets, applies joint factorization, reconstructs sample and feature embeddings,
+#' optionally projects test samples, computes a sample distance matrix, and returns cross-validation error statistics.
+#'
+#' @param tables A list of compositional matrices or data frames with features as rows and samples as columns.
+#' @param n_components Number of principal components to compute.
+#' @param max_iterations Maximum number of optimization iterations for OptSpace.
+#' @param test_samples Character vector of sample IDs to be projected into the ordination space.
+#' @param train_samples Character vector of sample IDs used to fit the ordination.
+#'
+#' @return A list with:
+#' \describe{
+#'   \item{ord_res}{An \code{OrdinationResults} object containing embeddings, loadings, and variance explained.}
+#'   \item{dist}{A \code{DistanceMatrix} object for sample embeddings.}
+#'   \item{cv_stats}{A data frame summarizing reconstruction error across iterations and tables.}
+#' }
+#'
+#' @keywords internal
+
+.joint_optspace_helper <- function(tables,
+                                   n_components,
+                                   max_iterations,
+                                   test_samples,
+                                   train_samples) {
+  #split and transpose training/test data per table
+  tables_split <- lapply(tables, function(tbl) {
+    list(t(tbl[, test_samples, drop = FALSE]),
+         t(tbl[, train_samples, drop = FALSE]))
+  })
+  
+  #format input for solver
+  tables_for_solver <- lapply(tables_split, function(pair) {
+    lapply(pair, as.matrix)
+  })
+  
+  #run joint OptSpace solver
+  opt_result <- .joint_optspace_solve(tables_for_solver,
+                                      n_components = n_components,
+                                      max_iter = max_iterations)
+  
+  U <- opt_result$U
+  S <- opt_result$S
+  V_list <- opt_result$V_list
+  dists <- opt_result$dists
+  
+  #assign row/column names to loadings
+  pc_names <- paste0("PC", seq_len(n_components))
+  
+  #combine feature loadings with table-derived row names
+  vjoint <- do.call(rbind, Map(function(tbl, V) {
+    rownames(V) <- rownames(tbl)
+    colnames(V) <- pc_names
+    V
+  }, tables, V_list))
```

Comment:
Bioconductor build checks require intendations of 4 spaces. 

Have you run the R & Bioconductor builds & checks?